File: fourth/fourth.mjs
import { readFileSync } from 'node:fs';

const lines = readFileSync('input.txt', { encoding: 'utf-8' }) // read day??.txt content
  .replace(/\r/g, '') // remove all \r characters to avoid issues on Windows
  .trim() // Remove starting/ending whitespace
  .split('\n');

function part1() {
  //do something here
  const result = lines.map((line) => {
    const [interval1, interval2] = line
      .split(',')
      .map((interval) => interval.split('-').map(Number))
      .sort((a, b) => {
        const oneSize = a[1] - a[0];
        const twoSize = b[1] - b[0];
        return twoSize - oneSize;
      });

    const oneFullContainsTwo =
      interval1[0] <= interval2[0] && interval1[1] >= interval2[1];

    return oneFullContainsTwo ? 1 : 0;
  });

  console.log(result.reduce((a, b) => a + b, 0));
}

function part2() {
  const result = lines.map((line) => {
    const [interval1, interval2] = line
      .split(',')
      .map((interval) => interval.split('-').map(Number))
      .sort((a, b) => {
        const oneSize = a[1] - a[0];
        const twoSize = b[1] - b[0];
        return twoSize - oneSize;
      });

    const overlap =
      interval1[1] >= interval2[0] && interval2[1] >= interval1[0];

    return overlap ? 1 : 0;
  });
  console.log(result.reduce((a, b) => a + b, 0));
}

part1();
part2();

File: fourth/fourth.py
with open('input.txt') as f:
    cnt = 0
    for line in f:
        line = line.split(",")
        first, second = line[0], line[1]
        f_1, f_2, s_1, s_2 = int(first.split(
            "-")[0]), int(first.split("-")[1]), int(second.split("-")[0]), int(second.split("-")[1])
        if (f_1 <= s_1 <= f_2) or (f_1 <= s_2 <= f_2) or (s_1 <= f_1 <= s_2) or (s_1 <= f_2 <= s_2):
            cnt += 1

print(cnt)
